src/main/java/dataMgmt/SessionStorage.java

```diff
@@ -99,50 +125,88 @@ public Project getProjectById(long projectId) {
      * @param projects the list of projects to set
      */
     public void setProjects(List<Project> projects) {
+        List<Project> oldValue = getProjects();
         this.projects = new HashMap<>();
         for (Project project : projects) {
             this.projects.put(project.getProjectID(), project);
         }
-        setChanged();
-        notifyObservers(projects);
+        notifyListeners(PROJECT_LIST, oldValue, getProjects());
     }
 
     /**
      * Add a project to the currently loaded projects. If one project already exists with the given id,
      * it will be overwritten.
      * @param project the project to add
      */
-    public void addProject(Project project) {
+    public void setProject(Project project) {
         this.projects.put(project.getProjectID(), project);
-        setChanged();
-        notifyObservers(projects);
+        notifyListeners(PROJECT_LIST, null, project);
+    }
+
+    /**
+     * Remove a project by its id, if it exists.
+     * @param id the id of the project to remove
+     */
+    public void removeProjectById(long id) {
+        Project old = this.projects.remove(id);
+        notifyListeners(PROJECT_LIST, old, null);
+    }
+
+    /**
+     * Set the given project id as a subscribed project, if it exists.
+     * @param id to set subscribed
+     */
+    public void setSubscribed(long id) {
+        if (this.projects.containsKey(id)) {
+            this.subscribedIds.add(id);
+            notifyListeners(SUBSCRIBED_PROJECTS, null, id);
+        }
+    }
+
+    /**
+     * Remove the given project id from the set of subscribed ids, if it exists.
+     * @param id to remove from subscribed set
+     */
+    public void setUnsubscribed(long id) {
+        if (this.projects.containsKey(id)) {
+            this.subscribedIds.remove(id);
+            notifyListeners(SUBSCRIBED_PROJECTS, id, null);
+        }
+    }
+
+    /**
+     * Get the set of subscribed ids
+     * @return set of subscribed ids
+     */
+    public Set<Long> getSubscribedIds() {
+        return this.subscribedIds;
     }
 
     /**
      * Change the status of a user-project key in the status map and notify observers of this change.
      * @param projectUserKey the user-project combo to change
      * @param status the new status
      */
-    // notifyObservers() is called to prompt the gui to change online displays
     public void changeProjectUserStatus(String projectUserKey, OnlineStatus status) {
+        AbstractMap.SimpleEntry<String, OnlineStatus> oldValue;
         synchronized (this) {
+            oldValue = new AbstractMap.SimpleEntry<>(projectUserKey, projectUserStatus.get(projectUserKey));
             projectUserStatus.put(projectUserKey, status);
         }
-        setChanged();
-        notifyObservers(projectUserStatus);
+        notifyListeners(PROJECT_USER_STATUS, oldValue,
+                new AbstractMap.SimpleEntry<>(projectUserKey, projectUserStatus.get(projectUserKey)));
```

Comment:
When we notify listeners, we send them the newValue and the oldValue of what was changed. In this case, the newValue and oldValue is not just one 'value', it is a pair. Therefore, we need to send them the 'pair' which is created by a literal map entry. I didn't change the implementation of the propertyChangeListener system to make this nicer because of it's very few [if not just one] occurrence(s).